controllers/hyperparametertuningjob/hyperparametertuningjob_controller.go

```diff
@@ -18,505 +18,339 @@ package hyperparametertuningjob
+				return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to delete hyperparameter tuning job"))
+			}
+			// Describe the new state of the job
+			if ctx.TuningJobDescription, err = ctx.SageMakerClient.DescribeHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to describe SageMaker hyperparameter tuning job"))
+			}
 		}
-		return RequeueAfterInterval(r.PollInterval, nil)
-	case sagemaker.HyperParameterTuningJobStatusCompleted, sagemaker.HyperParameterTuningJobStatusFailed, sagemaker.HyperParameterTuningJobStatusStopped:
-		log.Info("Job is in terminal state. Deleting spawned TrainingJobs and removing finalizer.")
-
-		// Delete all spawned TrainingJobs, retry if failure.
-		err := ctx.HpoTrainingJobSpawner.DeleteSpawnedTrainingJobs(ctx, ctx.Job)
-		if err != nil {
-			log.Info("Not all associated TrainingJobs jobs were deleted, will retry", "error", err)
-			return RequeueAfterInterval(r.PollInterval, nil)
+
+	case sagemaker.HyperParameterTuningJobStatusStopped, sagemaker.HyperParameterTuningJobStatusFailed, sagemaker.HyperParameterTuningJobStatusCompleted:
+		if controllers.HasDeletionTimestamp(ctx.TuningJob.ObjectMeta) {
+			return r.cleanupAndRemoveFinalizer(ctx)
 		}
 
-		return r.removeFinalizerAndUpdate(ctx)
+	case sagemaker.HyperParameterTuningJobStatusStopping:
+		break
+
 	default:
-		log.Info("Job is in unknown status")
-		return NoRequeue()
+		unknownStateError := errors.New(fmt.Sprintf("Unknown Tuning Job Status: %s", ctx.TuningJobDescription.HyperParameterTuningJobStatus))
+		return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, unknownStateError)
 	}
-}
 
-func (r *HyperparameterTuningJobReconciler) addFinalizerAndRequeue(ctx reconcileRequestContext) (ctrl.Result, error) {
-	ctx.Job.ObjectMeta.Finalizers = append(ctx.Job.ObjectMeta.Finalizers, SageMakerResourceFinalizerName)
-	ctx.Log.Info("Add finalizer and Requeue")
-	prevGeneration := ctx.Job.ObjectMeta.GetGeneration()
-	if err := r.Update(ctx, &ctx.Job); err != nil {
-		ctx.Log.Error(err, "Failed to add finalizer", "StatusUpdateError", ctx)
-		return RequeueIfError(err)
+	if err = r.updateStatus(ctx, string(ctx.TuningJobDescription.HyperParameterTuningJobStatus)); err != nil {
+		return err
 	}
 
-	return RequeueImmediatelyUnlessGenerationChanged(prevGeneration, ctx.Job.ObjectMeta.GetGeneration())
+	return nil
 }
 
-func (r *HyperparameterTuningJobReconciler) getHyperParameterTuningJobName(state hpojobv1.HyperparameterTuningJob) string {
-	return GetGeneratedJobName(state.ObjectMeta.GetUID(), state.ObjectMeta.GetName(), 32)
-}
+// Initialize fields on the context object which will be used later.
+func (r *Reconciler) initializeContext(ctx *reconcileRequestContext) error {
+	// Ensure we generate a new name and populate the spec if none was specified
+	if ctx.TuningJob.Spec.HyperParameterTuningJobName == nil {
```

Comment:
I can't respond to the earlier thread due to either my lack of understanding of how to use GitHub PR or GitHub PR has a bug.

> @cadedaniel
> cadedaniel 7 days ago Member
> 
> What if the name is empty but not nil?
> 
>  @cadedaniel
> cadedaniel 7 days ago Member
> This might be a bug in TrainingJob as well.
> 
>  @RedbackThomson
> RedbackThomson 3 days ago Author Member
> We decided that it would be better to do input validation (minimum length requirements) in the CRD spec, instead.
> 
>  @cadedaniel
> cadedaniel 3 days ago Member
> I don't see minimum length validation for HPO or TJ name. Can you include change for HPO and fix TJ?
> 
>  @RedbackThomson
> RedbackThomson 2 days ago Author Member
> I think all the CRD spec changes should be their own PR since there are probably other fields we should validate have non-empty validation.

We should improve validation in another PR, but we can't make the controller rely on validation that doesn't exist. The fix here is just adding MinLength validation to HPO and TJ. If you don't want to expand scope by adding validation, then this line should check for empty name.